### PR TITLE
routines with extended inputs/outputs and overrides

### PR DESCRIPTION
**Pull request by polomani**
Changes: 

- change [RoutineRunnerBase.cs](https://github.com/cognitedata/dotnet-simulator-utils/pull/110/files#diff-43eace63045f46b790d53c84abfca542e9d457315c6e6c2894c756bfd5121bd2) so it supports reading inputs from the new run/data endpoint as well as writing them back via callback.
- Make inputs overridable per simulation run
- Use SimulatorRoutineRevisionInfo type natively instead of the legacy SimulatorCalculation
- save inputs/outputs to time series only optionally if SaveTimeseriesExternalId property set
- change [RoutineImplementationBase.cs](https://github.com/cognitedata/dotnet-simulator-utils/pull/110/files#diff-c03ad250fea0bc2572b8bead4c85a3123e32808df62eccb47407450c20d24f3e) so it supports new inputs/outputs types

Other changes:
- removed the writeback of the model version to the timeseries (GetOrCreateSimulationModelVersion). This is now not in use because we have a better API that covers the same feature
- [SequencesExtensions.cs](https://github.com/cognitedata/dotnet-simulator-utils/pull/110/files#diff-ff0262dbd614a25645d18558f9c9c32b027810d20f4314b25221877e7c0e68b6) are removed for the same reason
- refactored the scheduler

TODO in next PRs:
- remove the use of `argType == outputTimeseries`, as well as `inputTimeSeries` and `inputConstant`. This is now sort of useless.

### CODE REVIEW

File: Cognite.Simulator.Utils/ConfigurationLibraryBase.cs [hunk 1]
```diff
@@ -175,88 +182,88 @@ protected virtual T StateFromRoutineRevision(SimulatorRoutineRevision routineRev
...
-                    Steps = s.Steps.Select((step, j) => new CalculationProcedureStep()
-                    {
-                        Step = j,
-                        Type = step.StepType,
-                        Arguments = step.Arguments
-                    })
-                }),
-                CreatedTime = routineRev.CreatedTime
-            };
-            localConfiguration = LocalConfigurationFromRoutine(simulationConfigurationWithRoutine);
```
Comment (polomani) [inline Cognite.Simulator.Utils/ConfigurationLibraryBase.cs:259]: the biggest change is here basically. we don't use `SimulationConfigurationWithRoutine` anymore

File: Cognite.Simulator.Utils/SimulationScheduler.cs [hunk 2]
```diff
@@ -135,44 +136,46 @@ public async Task Run(CancellationToken token)
...
-                RunType = "scheduled",
-                UserEmail = calcConfig.UserEmail
-            };
-        }
+        // private SimulationRunCreate CreateRunEvent(U calcState, V calcConfig)
+        // {
+        //     // return new SimulationEvent
+        //     // {
+        //     //     Calculation = calcConfig.RoutineExternalId,
+        //     //     Connector = _config.GetConnectorName(),
```
Comment (polomani) [inline Cognite.Simulator.Utils/SimulationScheduler.cs:172]: no need of this object either, we use SimulationRunCreate directly

File: Cognite.Simulator.Utils/CommonUtils.cs [hunk 3]
```diff
@@ -181,14 +189,18 @@ public static class SimulationUtils
...
+                throw new ArgumentException($"Logical check operator not recognized: {lcConfig.Operator}", nameof(lcConfig));
+            }
+            if (!lcConfig.Value.HasValue)
             {
-                throw new ArgumentException($"Logical check operator not recognized: {lcConfig.Check}", nameof(lcConfig));
+                throw new ArgumentException("Logical check value is missing", nameof(lcConfig));
             }
 
-            return DataSampling.LogicalCheck(ts, lcConfig.Value, op, validationRange.Max);
+            return DataSampling.LogicalCheck(ts, lcConfig.Value.Value, op, validationRange.Max); // todo defaults
```
Comment (polomani) [inline Cognite.Simulator.Utils/CommonUtils.cs:203]: does todo make sense anymore?

File: Cognite.Simulator.Utils/CommonUtils.cs [hunk 4]
```diff
@@ -112,38 +113,45 @@ public static class SimulationUtils
...
                 ssdMap = Detectors.SteadyStateDetector(
                     ssDps.ToTimeSeriesData(
                         config.DataSampling.Granularity,
-                        config.SteadyStateDetection.AggregateType.ToDataPointAggregate()),
-                    config.SteadyStateDetection.MinSectionSize,
-                    config.SteadyStateDetection.VarThreshold,
-                    config.SteadyStateDetection.SlopeThreshold);
+                        steadyStateDetection.Aggregate.ToDataPointAggregate()),
+                    steadyStateDetection.MinSectionSize ?? 0, // TODO what are the defaults?
+                    steadyStateDetection.VarThreshold ?? 0,
```
Comment (polomani) [inline Cognite.Simulator.Utils/CommonUtils.cs:153]: remove the todo

File: Cognite.Simulator.Extensions/Cognite.Simulator.Extensions.csproj [hunk 5]
```diff
@@ -22,6 +22,7 @@
   </ItemGroup> 
   <ItemGroup>
     <PackageReference Include="CogniteSdk" Version="3.26.1" />
```
Comment (polomani) [inline Cognite.Simulator.Extensions/Cognite.Simulator.Extensions.csproj:24]: update the version of the simutils

**Review by abdullah-cognite — APPROVED** [review]

**Pull request merged**